src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -672,23 +672,13 @@ public override void DidCompleteWithError (NSUrlSession session, NSUrlSessionTas
 
 			void SetResponse (InflightData inflight)
 			{
-				lock (inflight.Lock) {
-					if (inflight.ResponseSent)
-						return;
-
-					if (inflight.CancellationTokenSource.Token.IsCancellationRequested)
-						return;
-
-					if (inflight.CompletionSource.Task.IsCompleted)
-						return;
-
-					var httpResponse = inflight.Response;
+				if (inflight.CancellationTokenSource.Token.IsCancellationRequested)
+					return;
 
-					inflight.ResponseSent = true;
+				if (inflight.CompletionSource.Task.IsCompleted)
+					return;
 
-					// EVIL HACK: having TrySetResult inline was blocking the request from completing
```

Comment:
I don't understand why this hack was necessary in the first place, exactly what was blocking? If some other code is waiting for the CompletionSource to complete, and executes code when that happens, it's that code's responsibility to not block.

Reply:
I was not the person that wrote this. But in the tests, the lock + the TrySetResult blocks meaning that the request would never complete.

Reply:
There has been a few other threading fixes in the past. It's possible that the _original_ condition does not happen anymore, but it's also possible it's just harder to reproduce. How do you intend to test this ?

Reply:
It is already tested, all the tests in https://github.com/xamarin/xamarin-macios/blob/master/tests/monotouch-test/System.Net.Http/MessageHandlers.cs will timeout and fail if the call blocks. The blocking happens both on async and sync calls. There is no need to add a specific test for this.

Reply:
Our test suite has historically not been good to catch threading errors, like missing locks. I'm not so much afraid of the EVIL HACK removal than the rest.

Reply:
We are keeping the lock, so no change there. The lock was added if there is more than one thread accessing the inflight data, I don't think it will happen. The task.run was added due to https://github.com/xamarin/xamarin-macios/pull/8555#issuecomment-628114963 but moving the token creation to be async fixes the lock and the memory leak (in managed objects).

Reply:
The change is less risky than it was. Still, with recent changes, I don't think we have the correct setup to test this properly (which goes beyond unit testing). c.c. @chamons